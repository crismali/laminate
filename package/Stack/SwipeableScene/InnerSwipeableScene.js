import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { SwipeCatcher } from '../../SwipeCatcher'

export class InnerSwipeableScene extends Component {
  static defaultProps = {
    completeThreshold: 0.5
  }

  static propTypes = {
    children: PropTypes.any.isRequired,
    completeThreshold: PropTypes.number,
    history: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired,
    setAnimationValue: PropTypes.func.isRequired
  }

  handleMove = ({ percentComplete, velocity }) => {
    this.props.setAnimationValue(1 - percentComplete)
  }

  handleGestureRelease = ({ percentComplete, velocity }) => {
    const { completeThreshold, history, location } = this.props

    if (percentComplete > completeThreshold) {
      history.goBack()
    } else {
      history.replace(location)
    }
  }

  handleGestureTerminate = ({ percentComplete, velocity }) => {
    const { history, location } = this.props
    history.replace(location)
  }

  render () {
    return <SwipeCatcher
      {...this.props}
      onMove={this.handleMove}
      onGestureRelease={this.handleGestureRelease}
      onGestureTerminate={this.handleGestureTerminate}
    />
  }
}
