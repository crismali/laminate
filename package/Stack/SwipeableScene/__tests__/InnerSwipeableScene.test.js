/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { InnerSwipeableScene } from '../InnerSwipeableScene'

const Child = () => <div>a child</div>

describe('#handleMove', () => {
  it('calls set animation value', () => {
    const setAnimationValue = jest.fn()
    const subject = shallow(<InnerSwipeableScene setAnimationValue={setAnimationValue} history={{}} location={{}}>
      <Child />
    </InnerSwipeableScene>)
    subject.instance().handleMove({ percentComplete: 0.3 })
    expect(setAnimationValue).toHaveBeenCalledWith(0.7)
  })
})

describe('#handleGestureRelease', () => {
  it('goes back when it passes the threshold', () => {
    const history = { goBack: jest.fn(), replace: jest.fn() }
    const location = { the: 'location' }
    const subject = shallow(<InnerSwipeableScene completeThreshold={0.3} setAnimationValue={jest.fn()} history={history} location={location}>
      <Child />
    </InnerSwipeableScene>)
    subject.instance().handleGestureRelease({ percentComplete: 0.4 })
    expect(history.goBack).toHaveBeenCalled()
    expect(history.replace).not.toHaveBeenCalled()
  })

  it('replaces when it does not pass the threshold', () => {
    const history = { goBack: jest.fn(), replace: jest.fn() }
    const location = { the: 'location' }
    const subject = shallow(<InnerSwipeableScene completeThreshold={0.3} setAnimationValue={jest.fn()} history={history} location={location}>
      <Child />
    </InnerSwipeableScene>)
    subject.instance().handleGestureRelease({ percentComplete: 0.2 })
    expect(history.replace).toHaveBeenCalledWith(location)
    expect(history.goBack).not.toHaveBeenCalled()
  })
})

describe('#handleGestureTerminate', () => {
  it('replaces', () => {
    const history = { replace: jest.fn() }
    const location = { the: 'location' }
    const subject = shallow(<InnerSwipeableScene setAnimationValue={jest.fn()} history={history} location={location}>
      <Child />
    </InnerSwipeableScene>)
    subject.instance().handleGestureRelease({})
    expect(history.replace).toHaveBeenCalledWith(location)
  })
})

describe('#render', () => {
  it('renders properly', () => {
    const history = { replace: jest.fn() }
    const location = { the: 'location' }
    const subject = shallow(<InnerSwipeableScene fun='prop' setAnimationValue={jest.fn()} history={history} location={location}>
      <Child />
    </InnerSwipeableScene>)
    expect(subject).toMatchSnapshot()
  })
})
