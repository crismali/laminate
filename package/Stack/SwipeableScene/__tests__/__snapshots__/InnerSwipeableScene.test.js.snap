// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render renders properly 1`] = `
<SwipeCatcher
  completeThreshold={0.5}
  direction="leftToRight"
  fun="prop"
  history={
    Object {
      "replace": [Function],
    }
  }
  location={
    Object {
      "the": "location",
    }
  }
  onGestureRelease={[Function]}
  onGestureTerminate={[Function]}
  onMove={[Function]}
  setAnimationValue={[Function]}
>
  <Child />
</SwipeCatcher>
`;
